### PR TITLE
Introduce transaction domain tag

### PR DESCRIPTION
**Pull request by janezpodhostnik**
Closes: #61 

## Description

Added the transaction domain tag before signing the transaction. The transaction domain tag can be changed, but it has to be changed by all signing parties (if the tx is encoded/decoded when passed from one signer to another signer).

______

For contributor use:

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [standards mentioned here](https://github.com/onflow/flow-go-sdk/blob/master/CONTRIBUTING.md#styleguides).
- [ ] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: transaction.go [hunk 1]
```diff
@@ -74,17 +74,24 @@ type Transaction struct {
 
 	// PayloadSignatures is a list of signatures generated by the proposer and authorizer roles.
 	//
-	// A payload signature is generated over the inner portion of the transaction (payload).
+	// A payload signature is generated over the inner portion of the transaction (TransactionDomainTag + npayload).
```
Comment (turbolent) [inline transaction.go:77]:
```suggestion
	// A payload signature is generated over the inner portion of the transaction (TransactionDomainTag + payload).
```

**Comment by turbolent** [comment]
@tarakby can you please have a look here?

File: transaction.go [hunk 2]
```diff
@@ -74,17 +74,24 @@ type Transaction struct {
...
 	//
 	// You can find more information about transaction signatures here: https://docs.onflow.org/concepts/transaction-signing/#anatomy-of-a-transaction
 	EnvelopeSignatures []TransactionSignature
+
+	// TransactionDomainTag is prepended to the transaction message before it is signed.
+	//
+	// The transaction domain tag is not encoded together with the transaction. The default transaction domain tag is `paddedDomainTag("FLOW-V0.0-transaction")`
+	//
+	// You can find more information about transaction domain tag here: https://docs.onflow.org/concepts/transaction-signing/#anatomy-of-a-transaction
+	TransactionDomainTag []byte
```
Comment (tarakby) [inline transaction.go:94]: `TransactionDomainTag` is a protocol-wide constant, rather than a transaction-specific field. I don't think it should be added to the transaction structure. 

The constant can be used directly in `SignEnvelope` and `SignPayload`.

File: transaction.go [hunk 3]
```diff
@@ -304,7 +315,9 @@ func (t *Transaction) SignPayload(address Address, keyIndex int, signer crypto.S
 //
 // This function returns an error if the signature cannot be generated.
 func (t *Transaction) SignEnvelope(address Address, keyIndex int, signer crypto.Signer) error {
-	sig, err := signer.Sign(t.EnvelopeMessage())
+	message := t.EnvelopeMessage()
+	message = append(t.TransactionDomainTag, message...)
+	sig, err := signer.Sign(message)
```
Comment (tarakby) [inline transaction.go:320]: The comment doesn't belong to the code above:
Does the SDK also implement verifications for the payload and envelope signatures? If yes, then the tag needs to be prepended to the message in those functions as well.

File: transaction.go (hunk 2 above)
Comment (janezpodhostnik) [reply transaction.go:94]: the `TransactionDomainTag` has an associated version: `paddedDomainTag("FLOW-V0.0-transaction")`. Is there no chance that in the future if this version changes, someone will want to use the new version before the sdk is updated (or without updating the sdk)?

Comment (tarakby) [reply transaction.go:94]: It's still not clear to me what the versioning will be used for (versions of the transaction model, versions of Flow forks, ..). For now the FMV only supports a unique static domain tag when verifying signatures, and not a dynamic one read on the transaction. If we were to enable multiple version support, we could have a specific field for for the version in the transaction structure instead of the entire tag, and that field will be checked by the FVM (and collection nodes) for transaction validity.  

Btw, I think it's a good idea to make the SDK re-use the [tag used by the FVM ](https://github.com/onflow/flow-go/blob/master/model/flow/constants.go#L43).

Comment (janezpodhostnik) [reply transaction.go:94]: > It's still not clear to me what the versioning will be used for (versions of the transaction model, versions of Flow forks, ..). For now the FMV only supports a unique static domain tag when verifying signatures, and not a dynamic one read on the transaction. If we were to enable multiple version support, we could have a specific field for for the version in the transaction structure instead of the entire tag, and that field will be checked by the FVM (and collection nodes) for transaction validity.

I understand. I will change it as you propose.

> Btw, I think it's a good idea to make the SDK re-use the tag used by the FVM .

`flow-go-sdk` does not depend on flow-go. In fact flow-go depends on the `flow-go-sdk`. So If anything, I could change `flow-go` to use the tag defined in the sdk, but that does not seem like a good idea.

**Review by tarakby — APPROVED** [review]
🏄‍♂️

File: client/convert/protobuf.go [hunk 4]
```diff
@@ -444,7 +444,7 @@ func TransactionToMessage(t flow.Transaction) (*entities.Transaction, error) {
 
 func MessageToTransaction(m *entities.Transaction) (flow.Transaction, error) {
 	if m == nil {
-		return flow.Transaction{}, ErrEmptyMessage
+		return *flow.NewTransaction(), ErrEmptyMessage
```
Comment (psiemens) [inline client/convert/protobuf.go:447]: Why the change from struct literals to a constructor? Is that still needed now that the `TransactionDomainTag` field is removed?

File: transaction.go (hunk 3 above)
Comment (psiemens) [reply transaction.go:311]: No, the Go SDK doesn't have any transaction signature verification functionality.

File: client/convert/protobuf.go (hunk 4 above)
Comment (psiemens) [reply client/convert/protobuf.go:447]: If possible, I would change back just to keep this PR minimal.

**Review by psiemens — COMMENTED** [review]
Nice!

File: transaction.go (hunk 3 above)
Comment (tarakby) [reply transaction.go:311]: Thanks @Peter!
Do you know if there is a flow-go issue to mirror this SDK change in the verifications on the FVM side  ?

File: transaction.go (hunk 2 above)
Comment (tarakby) [reply transaction.go]: Hm right, the only dependency I know in the opposite way  is `flow-go-sdk/crypto` depending on `flow-go/crypto`. 
I'll add the change on `flow-go` then.

File: transaction.go (hunk 3 above)
Comment (janezpodhostnik) [reply transaction.go:311]: @tarakby this is the change in flow-go https://github.com/onflow/flow-go/pull/606
